### PR TITLE
SM-281: Secrets Manager Trash

### PR DESCRIPTION
**Pull request by coltonhurst**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR adds Trash support to Secrets Manager. This PR gives the API support for the following:
- Viewing secrets in the trash
- Permanently deleting secrets
- Restoring secrets

You can find the paired `clients` PR here: https://github.com/bitwarden/clients/pull/4730

## Code changes

* **EmptyTrashCommand.cs:** Add a command for emptying the trash (by secret id's)
* **RestoreTrashCommand.cs:** Add a command for restoring the trash (by secret id's)
* **SecretsManagerCollectionExtensions.cs:** Add DI support
* **SecretRepository.cs:** Add ability to get secrets in trash (by orgId & by orgId + secretIds)
* **TrashController.cs:** Add the TrashController for the new trash endpoints
* **BulkRestoreResponseModel.cs:** The bulk restore model, adds error support for when we have permission checks
* **IEmptyTrashCommand.cs:** Interface for the empty trash command
* **IRestoreTrashCommand.cs:** Interface for the restore trash command
* **ISecretRepository.cs:** Updates the SecretRepository interface to support the two new functions

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Trash/EmptyTrashCommand.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        var results = ids.Select(id =>
+        {
+            var secret = secrets.FirstOrDefault(secret => secret.Id == id);
+            if (secret == null)
+            {
+                throw new NotFoundException();
+            }
+            // TODO Once permissions are implemented add check for each secret here.
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Trash/EmptyTrashCommand.cs:36]: Only admins can access trash right, so we probably don't need this here. Which probably lets us remove this logic and just throw a 404 error if they attempt to delete an invalid secret.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Trash/RestoreTrashCommand.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            if (secret == null)
+            {
+                throw new NotFoundException();
+            }
+            // TODO Once permissions are implemented add check for each secret here.
+            else
+            {
+                return new Tuple<Secret, string>(secret, "");
+            }
+        }).ToList();
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Trash/RestoreTrashCommand.cs:41]: Same here, it's limited to admins only so we can remove the access check.

File: src/Api/SecretsManager/Controllers/TrashController.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+using Bit.Core.SecretsManager.Commands.Trash.Interfaces;
+using Bit.Core.SecretsManager.Repositories;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Bit.Api.SecretsManager.Controllers;
+
+[SecretsManager]
+[Authorize("secrets")]
+public class TrashController : Controller
```
Comment (Hinton) [inline src/Api/SecretsManager/Controllers/TrashController.cs:15]: Should we call this `SecretsTrashController`?

**Review by Hinton — COMMENTED** [review]
Looks good, some comments that we can remove the access check comments which lets us simplify the commands and responses to void.

**Review by Hinton — APPROVED** [review]
LGTM, @coltonhurst will review my code.

**Comment by coltonhurst** [comment]
@Hinton The code additions look good, thank you!

**Pull request merged**